tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -1454,38 +1454,164 @@ Status Converter::RenameAndMarkOutputTensors(
+                   "non-empty.";
+
+        VLOG(1) << "List of allowed tactics:";
+        VLOG(1) << absl::StrJoin(allowed_algorithm_ids, ", ");
 
+        VLOG(1) << "List of disallowed tactics:";
+        VLOG(1) << absl::StrJoin(disallowed_algorithm_ids, ", ");
+
+        VLOG(1) << "List of tactics common to both allowed and disallowed "
+                   "tactics:";
+        VLOG(1) << absl::StrJoin(intersection, ", ");
+
+        VLOG(1) << "Aborting.";
+        assert(0);
+      }
+    }
+  }
   // Returns value in [0, nbChoices] for a valid algorithm.
   int32_t selectAlgorithms(const nvinfer1::IAlgorithmContext& algoContext,
                            const nvinfer1::IAlgorithm* const* algoChoices,
-                           int32_t nbChoices,
-                           int32_t* selection) noexcept override {
+                           int32_t nbChoices, int32_t* selection) override {
+
+    int nbSelections = 0;
+
     // TensorRT always provides more than zero number of algorithms
     // in selectAlgorithms.
     assert(nbChoices > 0);
 
-    // making sure that the requested TRT algorithm ID doesn't go above the
-    // max value accepted.
-    selection[0] = std::min(algorithm_id_, nbChoices);
-    return 1;
+    absl::optional<int64_t> forced_algorithm_id  = absl::nullopt;
+
+    if (! allowed_algorithm_ids.empty()) {
+      assert(allowed_algorithm_ids.size() == 1);
+      forced_algorithm_id = *allowed_algorithm_ids.begin(0);
+      // Making sure that the requested TRT algorithm ID doesn't go above
+      // the max value accepted.
+      forced_algorithm_id = std::min(forced_algorithm_id.value(),
+                                     static_cast<int64_t>(nbChoices));
+
+      VLOG(1) << "Forcing TRT algorithm selection to: ID = "
+              << forced_algorithm_id.value();
+      selection[nbSelections++] = forced_algorithm_id.value();
+    }
+
+    auto forced_algorithm_id_is_banned = [](int64_t _algorithm_id){
+      VLOG(1) << "Impossible to force TensorRT tactic selection. All "
+              << "attempts have been rejected. Please select a different "
+              << "value for `TF_TRT_FIXED_ALGORITHM_ID`. Received: "
+              << _algorithm_id << ".";
+      assert(false);
+    };
+
+    // Disable all disallowed tactics
+    for (auto i = 0; i < nbChoices; i++) {
+      if (algoChoices[i] == nullptr) {
+        continue;
+      }
+
+      int64_t tacticId = algoChoices[i]->getAlgorithmVariant().getTactic();
+
+      if (disallowed_algorithm_ids.find(tacticId) !=
+          disallowed_algorithm_ids.end()) {
+        if (!forced_algorithm_id.has_value()) {
+          VLOG(1) << "Rejecting a disallowed tactic: " << tacticId
+                  << " for node " << algoContext.getName();
+          continue;
+        } else if (forced_algorithm_id.value() != i) {
+          continue;
+        }
+        forced_algorithm_id_is_banned(i);
```

Comment:
we can replace this with:
VLOG(1) << "Impossible to force TensorRT tactic selection. All "
              << "attempts have been rejected. Please select a different "
              << "value for `TF_TRT_FIXED_ALGORITHM_ID`. Received: "
              << id << ".";
      assert(false);

Reply:
I used the lambda function to avoid code duplication since I need to issue this error at two place in the code

Reply:
I am fine with keeping this lambda, if you DO NOT want the log messages look different for the two cases where it is used.

Reply:
I see your point, let me update that

Reply:
Fixed